documentation/directory-structure.md

```diff
@@ -30,13 +34,72 @@ You may be also interested in following:
 > The structure described in this article is strictly enforced. There exist some exceptions for historical reasons.
 > These exceptions are not allowed going forward.
 
+## Key concepts
+
+The directory structure is a reflection of few key concepts.
+
+A `service` is a set of operation endpoints (typically HTTP REST API endpoints) that **version uniformly**.
+
+A `service group` is a set of services that share common (ARM) **Resource Provider (RP) Namespace**, `RPNS`.
```

Comment:
```suggestion
A `service group` is a set of services that share a common (ARM) **Resource Provider (RP) Namespace**, `RPNS`.
```

Reply:
I think it's important to say that "services within a group can version independently of each other." Maybe put this in the "Uniform versioning" section? This independent versioning (with different docs, SDKs, ship timeframes, etc.) is the main motivation for having groups in the first place.

Reply:
Does this mean that, within a workgroup RP namespace, api-versions on the wire can be non-uniforrm across the resource namespace,  meaning that if service A uses api-version 2024-01-01 and service B uses api-version 2024-02-01, that service A's endpoints may not be available under api-version 2024-02-01?

Reply:
I think whoever initially came up with the name 'service group' could have clearly picked a better name. A service group as it stands today is a collection of resource types. That collection must version uniformly. A service is either a collection of resource types or a collection of service groups. We could possibly brainstorm on better names.

Reply:
Heres the hierarchy I am thinking of - 

Service - A collection of resource types or resource type groups. A service must be homogenous, so it must not have both resource types and resource type groups at the same level. 

Resource type groups - A collection of resource types. All resource types under a resource type group must version uniformly.

Resource type - An abstraction for a server side entity that is addressable through a url and operated on using HTTP methods.

Reply:
@rkmanda

Re `service group`. Consider these two distinct `services`:
- https://learn.microsoft.com/en-us/azure/aks/
- https://learn.microsoft.com/en-us/azure/kubernetes-fleet/

Both of them belong to the same `service group`. They share the same dir:
[`specification/containerservice/resource-manager/Microsoft.ContainerService`](https://github.com/Azure/azure-rest-api-specs/tree/main/specification/containerservice/resource-manager/Microsoft.ContainerService)

Seems to me this is a missing element from the hierarchy you proposed. 

Second thing: what is an effective distinction between a service that is a "collection of resource types" vs a "collection of resource type groups" ? How is this different from the point of view of our users? Is the end-user documentation organized differently, or the SDKs? Can you point me to an example? Here is a [resource provider list](https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/azure-services-resource-providers#match-resource-provider-to-service).

Because there is a distinction between `serviced and `service group` from the point of view of our users: 
- `service group` doesn't have to version uniformly, but the services it is composed of must
- Each `service` has its own corresponding SDK and documentation portal, but `service group` does not.

In your proposal, if the service is a collection of "resource type groups", then it would mean these **do not** have to version uniformly. Each `resource type group` has to version uniformly, but service is composed of multiple such groups. We cannot allow that. Entire service has to version uniformly, so a service **can not** be composed of **multiple** `resource type groups`.

Finally, you wrote:

>  A service is either a collection of resource types or a collection of service groups. 

This doesn't make sense to me, because I don't see how a service can be a collection of service groups. A service group is a collection of services, not the other way around (and with two levels of aggregation: the group itself is a collection, so a collection of service groups would be a collection of collection of services)

Reply:
@rkmanda Do we need the concept of a "resource type group"? Can we just say that a "service" is set of "resource types" that version uniformly? In other words, I think a "resource type group" is synonymous with a "service" and therefore we do not need 2 terms to represent a single abstraction.

Do you disagree with this thinking? If so, then I need to have the concept of "resource type group" explained to me.

Reply:
> I think whoever initially came up with the name 'service group' could have clearly picked a better name. A service group as it stands today is a collection of resource types. That collection must version uniformly. A service is either a collection of resource types or a collection of service groups. We could possibly brainstorm on better names.

@rkmanda I'm trying very hard to have the term "service" be what customers understand as the thing that versions, has docs/version, has an SDK/version, has a preview or stable version, etc. And then, we want to use the term "service" both within Microsoft and outside of Microsoft so we all use common terminology.

I'm trying to NOT burden customers with concepts of resource types, service groups, etc. These would be MS-internal only concepts used to give names to abstractions that WE talk abut but that customers have no need to talk about.

Reply:
> Does this mean that, within a workgroup RP namespace, api-versions on the wire can be non-uniforrm, meaning that if service A uses api-version 2024-01-01 and service B uses api-version 2024-02-01, that service A's endpoints may not be available under api-version 2024-02-01?

@markcowl Yes, there may be a "service-A" whose endpoints all use 2024-01-01 and another "service-B" whose endpoints all use 2024-02-1. But, I'm not sure what your use of "under" means in your question since endpoints are all flat.